### DISCUSSION TITLE
Why does not compiler compliain about non-initialized reference type members declared in structs?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/csharplang#6734] [Q&A] [answered]
```
struct Foo
{
	public byte[] Bar {get;set;}
}
```
Not complaining.

```
class Foo
{
	public byte[] Bar {get;set;}
}
```
Emits CS8618 warning.

Yes, I understand, that record can be created without init logic of constructor being called, but that means the default value is semantically invalid. I've declared `Bat` as `byte[]`, not as `byte[]?`, and I expect it to be not null.

Isn't it more correct to require all ref types in structs be nullable?

**Comment by CyrusNajmabadi** [comment, marked as answer]
NRT balances usefulness with accuracy. We deemed it too obnoxious to have to always nah Mark this as nullable and then have to deal with that at use sites. As most code doesn't run into default structures it was just onerous and massively increased the noise versus the signal. 

You can of course create an analyzer here trivially that would require you to make these all nullable and then have to deal with all the fallout.

**Reply by RikkiGibson** [reply]
One behavior people might find useful out of such an analyzer would be to tell you to make any such properties `required`, perhaps only if the type does not declare any constructors.

**Comment by TahirAhmadov** [comment]
Was it considered to track the state of struct as being default or not? Maybe, an attribute on method parameters and such:
```cs
struct Struct { public Struct() { this.Str = "abc"; } public string Str; }
void Foo(Struct s)
{
  int l = s.Str.Length; // NRT warning
}
void Bar([NotDefault] Struct s)
{
  int l = s.Str.Length; // no NRT warning
}

Struct s = default;
int l = s.Str.Length; // NRT warning
Foo(s); // no NRT warning
Bar(s); // NRT warning
s = new Struct();
int l2 = s.Str.Length; // no NRT warning
Foo(s); // no NRT warning
Bar(s); // no NRT warning
// etc.
```

**Reply by voroninp** [reply]
So every method which returns struct can also declare that it never returns default?

**Reply by TahirAhmadov** [reply]
Yes - I didn't list out all the places where such an attribute would be applied; I imagine fields, properties, method returns and parameters, and probably something else I'm not thinking of right now. 
It can also be made a keyword of some sort, such as, `not default`.
PS. Or a symbol: `void Bar(Struct! s) { ... }`

**Reply by voroninp** [reply]
It would also be nice if runtime provided a convenient way to detect that value I get is a default one. Because not always `default` is valid, and I want to have a guard check.

**Reply by theunrepentantgeek** [reply]
When does `default` not give you the default value for a struct? That's literally what it's spec'd to do. 

If it's not doing that, surely it's a bug.

**Reply by voroninp** [reply]
I mean, `default` is a way to bypass construction logic and field initializers if it was defined. That's why I'd like to know whether it's the case.

**Reply by HaloFour** [reply]
You can tell by comparing the value to `default`, although that only tells you if the internal state is zeroed-out, which could be a perfectly valid value (e.g. `0` for an `int`).

**Reply by RikkiGibson** [reply]
#5337 was recently considered and rejected by the nullability working group. I'm willing to consider approaches which are a little less complete and a little more practical, though.

It feels like tracking "default-ness" and actually warning when values might be default would require the ability to indicate that something is allowed to be default.

**Reply by voroninp** [reply]
@HaloFour , yes, but imagine `struct Either<TRight, TLeft>`,  comparing with `default` (especially in case `TRight` and `TLeft` are structs as well) may be not very performant thing to do.

**Reply by theunrepentantgeek** [reply]
From what I know of the way the compiler and JIT work, I think the comparison would be very performant indeed - and that nothing the runtime could do would be faster than `value == default`. 

A quick test with Benchmark.Net would confirm or refute this.

**Reply by RikkiGibson** [reply]
I don't think `==` is automatically defined on structs. It only exists if the user declares it (or if it is a special type).

**Reply by voroninp** [reply]
@HaloFour Also validity is a contextual thing, so if we treat constructor as a function which maps input values to struct, then the very fact that constructor was not explicitly called could mean a logical error, despite default value belonging to function domain.

**Reply by HaloFour** [reply]
@voroninp 

> Also validity is a contextual thing, so if we treat constructor as a function which maps input values to struct, then the very fact that constructor was not explicitly called could mean a logical error, despite default value belonging to function domain.

This is something that someone needs to keep in mind when they try to define such a type.  All structs have a default zeroed-out value, that is their nature, and you need to define such types to account for this.  `Either<TLeft, TRIght>` as a `struct` would have to account for `default` just like it would have to account for `null` if it were a `class` instead.

**Reply by theunrepentantgeek** [reply]
> I don't think == is automatically defined on structs

I stand corrected. You're right.

**Reply by voroninp** [reply]
@HaloFour , yep, I used private boolean field for that, but defaultness is why I tend to avoid records with behavior.
And that's why I have strict rule for enums: default value should never be a valid domain value.
